File: solutions/clojure/robot-simulator/1/src/robot_simulator.clj
(ns robot-simulator)

(defn robot [coord b]
  {:bearing b
   :coordinates coord})

(def bearings [:north :east :south :west :north])

(defn turn [b succession]
  (->> succession
       (partition 2 1)
       (filter #(= b (first %)))
       first
       second))

(defn turn-right [b]
  (turn b bearings))

(defn right [r]
  (robot (:coordinates r) (turn-right (:bearing r))))

(defn turn-left [b]
  (turn b (reverse bearings)))

(defn left [r]
  (robot (:coordinates r) (turn-left (:bearing r))))

(def displacement {:north {:x  0 :y  1}
                   :east  {:x  1 :y  0}
                   :south {:x  0 :y -1}
                   :west  {:x -1 :y  0}})

(defn advance [rob]
  (let [b (:bearing rob)
        disp (get displacement b)
        rob-coord (:coordinates rob)
        next-coord {:x (+ (:x disp) (:x rob-coord))
                    :y (+ (:y disp) (:y rob-coord))}]
    (robot next-coord b)))

(def commands {\R right
               \L left
               \A advance})

(defn simulate [command-str rob]
  (reduce 
   (fn [r c] ((get commands c) r))
   rob 
   command-str))

File: solutions/cpp/robot-simulator/1/robot_simulator.cpp
#include "robot_simulator.h"

namespace robot_simulator {
    Robot::Robot () {
        position = std::make_pair(0, 0);
        bearing = Bearing::NORTH;
    }
    Robot::Robot(std::pair<int, int> position, Bearing bearing)
        : position{position}, bearing{bearing} {}

    std::pair<int, int> Robot::get_position() const {
        return position;
    }

    Bearing Robot::get_bearing() const {
        return bearing;
    }

    void Robot::turn_left() {
        bearing = static_cast<Bearing>((static_cast<int>(bearing) + 4 - 1) % 4);
    }

    void Robot::turn_right() {
        bearing = static_cast<Bearing>((static_cast<int>(bearing) + 1) % 4);
    }

    void Robot::advance() {
        switch (bearing)
        {
        case Bearing::NORTH:
            position.second++;
            break;
        case Bearing::EAST:
            position.first++;
            break; 
        case Bearing::SOUTH:
            position.second--;
            break;
        case Bearing::WEST:
            position.first--;
            break;
        default:
            break;
        }
    }

    void Robot::execute_sequence(const std::string& sequence) {
        for (auto c : sequence) {
            switch (c)
            {
            case 'L':
                turn_left();
                break;
            case 'R':
                turn_right();
                break;
            case 'A':
                advance();
                break;
            default:
                break;
            }
        }
    }
}  // namespace robot_simulator
